### ISSUE TITLE
BoxCOCOMetrics don't support Ragged inputs

### ISSUE THREAD

**Issue by hugopi** [keras-team/keras-cv#1819]
Hi KerasCV ! 

I am using your RetinaNet Model and i would like to use coco metrics. When I train the model without specifying the metrics no problems. However when I precise the coco metric the following error is yield after the first epoch : 

Epoch 1: 
  `19/Unknown - 54s 384ms/step - loss: 1.3833 - box_loss: 0.4004 - classification_loss: 0.9828 - MaP: 0.0000e+00 - MaP@[IoU=50]: 0.0000e+00 - MaP@[IoU=75]: 0.0000e+00 - MaP@[area=small]: 0.0000e+00 - MaP@[area=medium]: 0.0000e+00 - MaP@[area=large]: 0.0000e+00 - Recall@[max_detections=1]: 0.0000e+00 - Recall@[max_detections=10]: 0.0000e+00 - Recall@[max_detections=100]: 0.0000e+00 - Recall@[area=small]: 0.0000e+00 - Recall@[area=medium]: 0.0000e+00 - Recall@[area=large]: 0.0000e+00 - percent_boxes_matched_with_anchor: 0.8336`

```
Traceback (most recent call last):
  File "/home/hugo/miniconda3/envs/sinbad/lib/python3.10/site-packages/tensorflow/python/ops/script_ops.py", line 269, in __call__
    return func(device, token, args)
  File "/home/hugo/miniconda3/envs/sinbad/lib/python3.10/site-packages/tensorflow/python/ops/script_ops.py", line 147, in __call__
    outputs = self._call(device, args)
  File "/home/hugo/miniconda3/envs/sinbad/lib/python3.10/site-packages/tensorflow/python/ops/script_ops.py", line 154, in _call
    ret = self._func(*args)
  File "/home/hugo/miniconda3/envs/sinbad/lib/python3.10/site-packages/tensorflow/python/ops/script_ops.py", line 813, in wrapped_func
    out = func(*structured_inp)
  File "/home/hugo/miniconda3/envs/sinbad/lib/python3.10/site-packages/tensorflow/python/autograph/impl/api.py", line 642, in wrapper
    return func(*args, **kwargs)
  File "/home/hugo/miniconda3/envs/sinbad/lib/python3.10/site-packages/keras_cv/metrics/object_detection/box_coco_metrics.py", line 167, in update_state_on_cpu
    return obj_update_state(y_true, y_pred, sample_weight)
  File "/home/hugo/miniconda3/envs/sinbad/lib/python3.10/site-packages/keras_cv/metrics/object_detection/box_coco_metrics.py", line 233, in update_state
    self._cached_result = self._compute_result()
  File "/home/hugo/miniconda3/envs/sinbad/lib/python3.10/site-packages/keras_cv/metrics/object_detection/box_coco_metrics.py", line 250, in _compute_result
    metrics = compute_pycocotools_metric(
  File "/home/hugo/miniconda3/envs/sinbad/lib/python3.10/site-packages/keras_cv/metrics/object_detection/box_coco_metrics.py", line 301, in compute_pycocotools_metric
    return keras_cv.metrics.coco.compute_pycoco_metrics(
  File "/home/hugo/miniconda3/envs/sinbad/lib/python3.10/site-packages/keras_cv/metrics/coco/pycoco_wrapper.py", line 225, in compute_pycoco_metrics
    gt_dataset = _convert_groundtruths_to_coco_dataset(groundtruths)
  File "/home/hugo/miniconda3/envs/sinbad/lib/python3.10/site-packages/keras_cv/metrics/coco/pycoco_wrapper.py", line 155, in _convert_groundtruths_to_coco_dataset
    max_num_instances = groundtruths["classes"][i].shape[1]
IndexError: tuple index out of range
	 [[{{node EagerPyFunc_1}}]] [Op:__inference_train_function_17573]
 ```
 
 I give you a link to a gist,  I tried to reproduce the error(maybe few modifications to do)  :
 
 https://gist.github.com/hugopi/a3d168046df760b601fdae3af9d12607
 
 Thank you for your support ! 
 
 Regards
 
 Hugo PIRES

**Comment by ianstenbit** [comment]
@hugopi it looks like the shape of your inputs doesn't include a batch size -- is that correct?

In this case, the metrics are expecting labels with the shape `[batch_size, num_boxes_per_batch]`, but it looks like the shape of your labels is `[None]` (presumably this represents `[batch_size]`. You'll need an extra dimension for `num_boxes_per_batch`. (The same is true for the boxes themselves, it looks like)

**Comment by hugopi** [comment]
Hi @ianstenbit ! 

Thanks for your answer but i need to clarify some points to make you understand my code i think. 

If on the surface my inputs seems to have a missing dimension for batch size due to the yield of my generator, this dimension is actually add with `padded_batch()`. 

You may be wonder why `padded_batch()` instead of `ragged_batch()`? Because I am using tensorflow 2.11 and `ragged_batch()` is a feature from tensorflow 2.12. So I create a padded batch dataset that i map to ragged with the `ragged_dataset()` function. 


This is the appearance of my real values in a dataset with `batch_size=8` : 

```
>>>train_dataset.element_spec
(TensorSpec(shape=(8, 512, 512, 3), dtype=tf.float32, name=None), {'boxes': RaggedTensorSpec(TensorShape([8, None, 4]), tf.float32, 1, tf.int64), 'classes': RaggedTensorSpec(TensorShape([8, None]), tf.float32, 1, tf.int64)})
```
`None` dimension representing number of detection in one element of the batch. It's `None` because each element of the batch could have a  different number of detection.

```
>>>next(iter(train_dataset))

(<tf.Tensor: shape=(8, 512, 512, 3), dtype=float32, numpy=
array([[[[0.22794858, 0.02173912, 0.22794858],
         [0.13921069, 0.02173912, 0.13921069],
         [0.07489516, 0.02173912, 0.07489516],
         ...,
         [0.12426693, 0.02173912, 0.12426693],
         [0.13279389, 0.02173912, 0.13279389],
         [0.15347545, 0.02173912, 0.15347545]],
        [[0.23022266, 0.02173912, 0.23022266],
         [0.15261193, 0.02173912, 0.15261193],
         [0.09595545, 0.02173912, 0.09595545],
         ...,
         [0.09794039, 0.02173912, 0.09794039],
         [0.17421728, 0.02173912, 0.17421728],
         [0.24660785, 0.02173912, 0.24660785]],
        [[0.22908357, 0.02173912, 0.22908357],
         [0.16956106, 0.02173912, 0.16956106],
         [0.12426693, 0.02173912, 0.12426693],
         ...,
         [0.08325624, 0.02173912, 0.08325624],
         [0.1937029 , 0.02173912, 0.1937029 ],
         [0.2968543 , 0.02173912, 0.2968543 ]],
        ...,
        [[0.02680415, 0.02173912, 0.02680415],
         [0.02437472, 0.02173912, 0.02437472],
         [0.03012604, 0.02173912, 0.03012604],
         ...,
        [...]
         ...,
         [0.9782608 , 0.02173912, 0.9782608 ],
         [0.9782608 , 0.02173912, 0.9782608 ],
         [0.9782608 , 0.02173912, 0.9782608 ]],
        [[0.88584316, 0.02173912, 0.88584316],
         [0.6117155 , 0.02173912, 0.6117155 ],
         [0.6360084 , 0.02173912, 0.6360084 ],
         ...,
         [0.9782608 , 0.02173912, 0.9782608 ],
         [0.78946066, 0.02173912, 0.78946066],
         [0.6478084 , 0.02173912, 0.6478084 ]],
        ...,
        [[0.6295348 , 0.02173912, 0.6295348 ],
         [0.72861284, 0.02173912, 0.72861284],
         [0.7831771 , 0.02173912, 0.7831771 ],
         ...,
         [0.5030216 , 0.02173912, 0.5030216 ],
         [0.559709  , 0.02173912, 0.559709  ],
         [0.6218429 , 0.02173912, 0.6218429 ]],
        [[0.8464153 , 0.02173912, 0.8464153 ],
         [0.9509264 , 0.02173912, 0.9509264 ],
         [0.69837785, 0.02173912, 0.69837785],
         ...,
         [0.5980263 , 0.02173912, 0.5980263 ],
         [0.5246342 , 0.02173912, 0.5246342 ],
         [0.5213425 , 0.02173912, 0.5213425 ]],
        [[0.9782608 , 0.02173912, 0.9782608 ],
         [0.9782608 , 0.02173912, 0.9782608 ],
         [0.6231193 , 0.02173912, 0.6231193 ],
         ...,
         [0.83307904, 0.02173912, 0.83307904],
         [0.5690824 , 0.02173912, 0.5690824 ],
         [0.46206063, 0.02173912, 0.46206063]]]], dtype=float32)>, 
         {'boxes': <tf.RaggedTensor [[[484.0, 293.0, 12.0, 29.0]], 
                                                      [[359.0, 76.0, 10.0, 13.0]],
                                                      [[197.0, 238.0, 10.0, 10.0]], 
                                                      [[43.0, 91.0, 10.0, 10.0],[349.0, 323.0, 10.0, 10.0]],
                                                      [[276.0, 345.0, 8.0, 10.0]], [[474.0, 82.0, 10.0, 10.0]],
                                                      [[137.0, 202.0, 6.0, 6.0],[137.0, 192.0, 2.0, 6.0],[468.0, 143.0, 12.0, 12.0]], 
                                                      [[281.0, 188.0, 8.0, 12.0],[211.0, 359.0, 8.0, 14.0],[324.0, 340.0, 22.0, 20.0]]]>, 
          'classes': <tf.RaggedTensor [[1.0], [0.0], [1.0], [1.0, 1.0], [1.0], [1.0], [1.0, 1.0, 1.0],[1.0, 1.0, 1.0]]>})
```

Is this the correct input shapes ? For the moment it's the only way I succeed to feed the model.

Concerning the `num_boxes_per_batch` dimension  you are talking about, is it number of detection per sample or per batch ? 
My inputs boxes shape : `'boxes': RaggedTensorSpec(TensorShape([8, None, 4])` ( `None` = `num_detection_in sample`, variable)

i.e Does  `num_boxes_per_batch` correspond to my `None` dimension or I have to add another dimension like this : 
`'boxes': RaggedTensorSpec(TensorShape([8, num_boxes_per_batch, num_detection_in sample, 4])`

Thanks a lot in advance for all your advice 

Regards 

Hugo PIRES

**Comment by james77777778** [comment]
Hi @hugopi 

I'm currently exploring object detection with KerasCV as well, and I have successfully conducted several experiments.

In tf2.11.0, you can use:
https://github.com/keras-team/keras-cv/blob/4a82e2261581c3655d1fdd68b732e3867a513bd9/examples/training/object_detection/pascal_voc/retinanet.py#L145-L147

When working with `keras_cv.metrics.BoxCOCOMetrics`, it MAY be necessary to convert the data into a dense tensor (I'm unsure about it, but it dose run):
https://github.com/keras-team/keras-cv/blob/4a82e2261581c3655d1fdd68b732e3867a513bd9/examples/training/object_detection/pascal_voc/retinanet.py#L151-L154

Hope this help

**Comment by hugopi** [comment]
Thank you @james77777778 ! 

For the ragged_batch dataset, yes I found the experimental methods in the week, I have quite the same pipeline.

Concerning keras_cv.metrics.BoxCOCOMetrics, i will try, thanks for the tip.

**Comment by hugopi** [comment]
Hi @ianstenbit and @james77777778,

Thanks for helping me resolve the problem successfully.

After conducting further experiments, it became evident that @james77777778 was indeed correct. I realized that I needed to provide dense inputs for cocometrics to yield accurate results. In the Keras cv tutorial, it was mentioned to skip the step of converting bounding boxes to dense if not using TPU. Being on a GPU, I naturally omitted this step. However, when attempting to utilize cocometrics, the conversion was necessary.

By incorporating dense inputs, I am now able to use coco metrics effectively during training or testing, without encountering any errors.

Once again, thank you, @ianstenbit and @james77777778, for your exceptional support.

Warmest regards,

Hugo PIRES
